machineconfiguration/v1alpha1/types_machineconfignode.go

```diff
@@ -0,0 +1,167 @@
+	// This value should match the desired version unless an upgrade is in progress.
+	// Must be a lowercase RFC-1123 hostname (https://tools.ietf.org/html/rfc1123)
+	// It may consist of only alphanumeric characters, hyphens (-) and periods (.)
+	// and must be at most 253 characters in length.
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +optional
+	Current string `json:"current"`
+	// desired is the MachineConfig the node wants to upgrade to.
+	// The machine config operator sets this value once it has
+	// generated a new machine config for the node to upgrade to.
+	// Must be a lowercase RFC-1123 hostname (https://tools.ietf.org/html/rfc1123)
+	// It may consist of only alphanumeric characters, hyphens (-) and periods (.)
+	// and must be at most 253 characters in length.
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +kubebuilder:validation:Required
+	Desired string `json:"desired"`
+}
+
+// MachineConfigNodeSpecMachineConfigVersion holds the desired config version for the current observed machine config node.
+// When Current != Desired; the MachineConfigOperator is in an upgrade phase and the machine config node will
+// take account of upgrade related events. Otherwise they will be ignored given that certain operations
+// happen both during the MCO's upgrade mode and the daily operations mode.
+type MachineConfigNodeSpecMachineConfigVersion struct {
+	// desired is the name of the machine config that the the node should be upgraded to.
+	// This value is set when the machine config pool generates a new version of its rendered configuration.
+	// When this value is changed, the machine config daemon starts the node upgrade process.
+	// If the transition of this value is accepted, the new desired version will be updated in the .status.configVersion.desired field
+	// and the upgrade will proceed.
+	// Must be a lowercase RFC-1123 hostname (https://tools.ietf.org/html/rfc1123)
+	// It may consist of only alphanumeric characters, hyphens (-) and period s(.)
+	// and must be at most 253 characters in length.
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +kubebuilder:validation:Required
+	Desired string `json:"desired"`
+}
+
+// StateProgress is each possible state for each possible MachineConfigNodeType
+// UpgradeProgression Kind will only use the "MachinConfigPoolUpdate..." types for example
+type StateProgress string
+
+const (
+	// MachineConfigNodeUpdatePrepared describes a machine that is preparing in the daemon to trigger an update
```

Comment:
Anything we can do to help people understand that these individual conditions are not yet set in stone?

Reply:
I think the idea of v1alpha1 alongside techpreview insinuates that, right? @JoelSpeed WDYT?

Reply:
With the addition of the proviso above the type definition I think we are ok here for now, this is tech preview, there is an expectation that there can be movement. In theory no one will use these except MCO until we release 4.15 anyway so I think we should be safe to include them and chop/change later